### ISSUE TITLE
Scenarios

### ISSUE THREAD

**Issue by dimitri-rusin** [dimitri-rusin/oll_onemax#14]
- lambdas: discrete action space [1, 2, 3, 4, 5, 6, 7, 8] VERSUS [2, 4, 6, 7, 8] VERSUS - lambdas: continuous action space: 1 - 8.
- DDQN versus PPO
- gap_to_optimality: 0.95 VERSUS 0.8 VERSUS 0.7 VERSUS 0.6 VERSUS 0.5

**Comment by ndangtt** [comment]
1. Problem sizes: 
- 500, 1000, 2000, 3000
- [optional] 10, 50, 100, 200

2. Action space: 
- Discrete: 
    - 100: [1, 10, 40, 60]
    - 1000: [x, x, x, x]
    - 3000: [x, x, x, x]
- Continuous: [1..n]

3. Initial solution:
- random 
- [optional] 80%, 95%


4. RL algorithm:
- DDQN (sb3, discrete action space)
- PPO (sb3, discrete & cont action space)
- [optional] Ray


**Next steps:**
- Fill in the holes: talk with Carola
- Define a default hyper-parameter setting for DDQN & PPO
- Define an RL training budget (for each problem size?)
- Launch experiments

**Comment by dimitri-rusin** [comment]
This is the experiment setting that's currently being evaluated with 10 random seeds, each at most for 7 hours, on Cirrus.

These are some general parameters:
```
dimensionality: 80
lambdas: [1, 2, 4, 8, 16, 32]
num_evaluation_episodes: 150
num_timesteps_per_evaluation: 400
closeness_to_optimum: 0.5
reward_type: EVALUATIONS_PLUS_FITNESS
state_type: ONE_HOT_ENCODED
num_environments: 16
max_training_timesteps: 2_000_000
```

These are PPO-specific parameters:
```
ppo:
  batch_size: 64
  clip_range: 0.2
  device: auto
  ent_coef: 0.0
  gae_lambda: 0.95
  gamma: 0.99
  learning_rate: 0.0003
  max_grad_norm: 0.5
  n_epochs: 10
  n_steps: 2_048
  policy: MlpPolicy
  sde_sample_freq: -1
  stats_window_size: 100
  verbose: 0
  vf_coef: 0.5
```

**Comment by ndangtt** [comment]
**Experiment setting:**
Problem sizes: **100**
num_evaluation_episodes: **500**
num_timesteps_per_evaluation: **2048**
#RL repetitions: **5 seeds**
#environments: **1**


**PPO hyper-parameters:**
n_steps: **512**

**Comment by ndangtt** [comment]
**Visualisation:**

- Visualise the training progress of each seed: 
   + performance of the baseline and learnt policy: mean and std.
   + marking points where the learnt policy reach 25%std of the mean of the baseline (as green points).

- Visualise results across multiple seeds as box-plots. Metrics to consider: AUC and hitting ratios.

- Visualise the learnt policy (x-axis: f(x), y-axis: lambda)

**Comment by ndangtt** [comment]
**Next steps:**

- Visualisation:
     - AUC + hitting_ratio boxplots
     - green marking of hitting points
     - baseline policy in the policy plot

- Results for 100, 200, 500
    - One-hot encoding
    - Scalar encoding
    
- DDQN with discrete action space

- PPO with continuous action space


- Compare pure Rust with Python/Rust policy evaluation: average number of timesteps and function evaluations. 500 runs.
- Check literature on reward scaling for PPO.

**Comment by dimitri-rusin** [comment]
Hi Nguyen, I have run  some preliminary experiments with reward scaling. I haven't done a literature review yet, but instead this is how the reward is computed
```py
reward = -num_function_evaluations_of_this_step + (self.current_fitness - prior_fitness)
reward /= self.dimensionality # for reward scaling
```

And these are the results (changing the dimensionality and the initial fitness):
http://178.128.240.117:8050/

**Comment by ndangtt** [comment]
hi @dimitri-rusin, thank you for the updates, the results look quite interesting!

From my understanding when looking at the plots in the link, scaling rewards by `n` (problem size) seems to help learning. Some interesting observations:

**For `n=100`**: 

The hitting ratio increases significantly for initial fitness of 0.5, but not very significant for initial fitness of 0.85. This is somewhat surprising to me. Do you have any insights on that?

Also I have a question: how do you calculate the green points? In our GECCO2022 paper, they're marked as green when they're within `mean(theory_based) +/- 25%*std(theory_based)`, is the same formula applied in your plots? Below is a fragment of `n=100, initial=0.85, with reward scaling`, the green points look like they're above the 25%std of `theory_based`:
![image](https://github.com/user-attachments/assets/c2b7402f-9551-44d7-82b8-9a0ea3dc80ee)

**For `n=200` and `n=500`:**

Reward scaling still seems to help but the increase in hitting ratio is not very significant. I'm wondering if increasing the scaling factor may help, e.g., `k*n` instead of `n`, where `k>1`. Deciding `k` can be tricky. One option is to calculate the running average/std of the total rewards (per episode) and use that to normalise the rewards on-the-fly (it's a common technique in RL). Looking at sb3's PPO implementation, I can see that they already implement it, see:

parameter `normalize_advantage`: https://stable-baselines3.readthedocs.io/en/master/modules/ppo.html
and their source code: https://stable-baselines3.readthedocs.io/en/master/_modules/stable_baselines3/ppo/ppo.html#PPO
```
if self.normalize_advantage and len(advantages) > 1:
                    advantages = (advantages - advantages.mean()) / (advantages.std() + 1e-8)
```

So in principle, if PPO default hyper-parameter setting is used (`normalize_advantage=True`), the advantage function should already be normalised. So I'm not sure why reward scaling still has a strong impact here.

sb3's PPO does provide some statistics output to the screen during the learning, do you still keep them? Perhaps they would provide some useful information and we could look at them together in the next meeting :)

**Comment by ndangtt** [comment]
In general, reward scaling seems to have a significant impact, but the scaling factor could be adjusted better. We want to look into this further.

TODO:

- Re-running a few PPO experiments with `verbose` and `tensorboard_log` activated. See why reward scaling has a positive impact even though PPO's advantage function is already normalised by default.

- Re-running a few PPO experiments with: `reward = reward / mean_fitness(theory_base)`

- Boxplots for AUC: calculate one AUC value for each run and then produce boxplots using those values. Same for the hitting ratio.

- Have a read at SB3 tips and tricks: https://stable-baselines3.readthedocs.io/en/master/guide/rl_tips.html
      + Some important points: normalise the observation space by `n`, scaling the action space to `[-1,1]` for continuous action space.

- PPO experiments on continuous action space, remember to try scaling the action space to `[-1,1]` as suggested in SB3's documentation.

- (Nguyen) Modify sb3's DQN so that it becomes DDQN.

**Comment by dimitri-rusin** [comment]
- Experiment to show that Python code is faster than Rust code in training.
- Which settings of PPO and which settings of the benchmark itself have we tried?

**Comment by dimitri-rusin** [comment]
The following are the parameters which we controlled in our experiments so far:

**Parameters of the OneMax landscape:**
```
dimensionality: [80, 100, 200, 500, 1_000]
closeness_to_optimum: [0.5, 0.85]
```

**Parameters of the (1 + (lambda, lambda)) genetic algorithm:**
```
lambdas: [[1, 2, 4, 8, 16, 32], [1, 2, 4, 8, 16, 32, 64], [1, 2, 4, 8, 16, 32, 64, 128]]
reward_type: [EVALUATIONS_PLUS_FITNESS, ONLY_EVALUATIONS]
state_type: [ONE_HOT_ENCODED, SCALAR_ENCODED]
random_seed: "list(range(100, 103))"
```

**Parameters of the Proximal Policy Optimization (PPO) reinforcement learning algorithm:**
```
max_training_timesteps: [1_000_000, 6_000_000]
num_environments: [1]
num_evaluation_episodes: [500] # Policy evaluation: How many episodes do we run to estimate the expected runtime of a generated policy?
num_timesteps_per_evaluation: [2_048] # How many RL steps do we run between two policy evaluations?
batch_size: [64]
clip_range: [0.2]
device: [auto]
ent_coef: [0.0]
gae_lambda: [0.95]
gamma: [0.99]
learning_rate: [0.0003]
max_grad_norm: [0.5]
n_epochs: [10]
n_steps: [2_048]
policy: [MlpPolicy]
sde_sample_freq: [-1]
stats_window_size: [100]
verbose: [0]
vf_coef: [0.5]
random_seed: "list(range(100, 103))"
```

The values that you see above are some of the values that we tried.

I don't have the outcome of experiments right now. But at least these are the parameters and some of their values that are of interest to us.

Of course, these are just the parameter names. Tai might be able to guess their meanings. Otherwise, what is missing here are answers to the following questions:

1. How does the (1 + (lambda, lambda)) genetic algorithm work with its parameters? And what do the parameters change theoretically?
2. How the PPO algorithm work with their parameters? And what do the parameters mean theoretically?
3. You probably know the OneMax problem :)
